### ISSUE TITLE
consider referencing the manager `cudd.BDD` in `cudd.Function.manager`

### ISSUE THREAD

**Issue by johnyf** [johnyf/dd#13]
Currently, the attribute [`dd.autoref.Function.bdd`](https://github.com/johnyf/dd/blob/e83f0a32705e6321c1e72aec7fdcd0d00559123b/dd/autoref.py#L217) points to the "low-level" manager [`dd.bdd.BDD`](https://github.com/johnyf/dd/blob/e83f0a32705e6321c1e72aec7fdcd0d00559123b/dd/bdd.py#L60), as indicated by [`dd.autoref.BDD._wrap`](https://github.com/johnyf/dd/blob/e83f0a32705e6321c1e72aec7fdcd0d00559123b/dd/autoref.py#L54).
Similarly, the attribute [`dd.cudd.Function.manager`](https://github.com/johnyf/dd/blob/e83f0a32705e6321c1e72aec7fdcd0d00559123b/dd/cudd.pyx#L1112) points to the "low-level" manager [`cudd/cudd.h/DdManager`](https://github.com/johnyf/cudd/blob/80c9396b7efcb24c33868aeffb89a557af0dc356/cudd/cudd/cudd.h#L281).
(Also, note the non-uniformity of the attribute name.)

If I recall correctly, the reason was that [flat is better than nested](https://www.python.org/dev/peps/pep-0020/). In other words, if `Function.manager` pointed to the associated python object `dd.cudd.BDD`, then accessing the associated `DdManager` would require reading the corresponding attribute of `Function.manager`. Besides, this is less efficient.

However, not referencing the python manager object within the `Function` object has some undesired and clumsy consequences:

- having to `del` each node defined, in order to avoid assertion errors upon exit, or even segfaults (depending on the underlying C package)

- need two lines to initialize a node, for example
  ```python
  f = Function()
  f.init(self.manager, u)
  ```
  where `self.manager` is a low-level manager.

An alternative worth considering is to reference the python object of the corresponding manager within the `Function` instance. Possibly, also retrieve the underlying "low-level" manager in `Function.init` and assign its value to some other attribute of `Function`. This will ensure less clutter in user code. (Unless one writes `del` to explicitly delete nodes to dereference them, which is an optimization.)

It is never the case that a `DdManager` isn't associated to an existing `BDD` instance, so the proposed approach doesn't introduce any restrictions.

## Initialization

Either way, the initialization of a `Function` costs the same, because the CUDD manager is obtained after access to an attribute (either `BDD.manager` or `Function.manager`). In other words, the current

```python
f = Function()
f.init(self.manager, u)
```

will be of equal cost as

```python
f = Function()
f.init(self, u)
```

because the attribute `self.manager` that is accessed within the caller in the first approach will be accessed within `init` in the second approach.
Besides, this cost should be negligible, compared to the bulk of the runtime, which is consumed within CUDD BDD operations.
The `init` method is not visible outside Cython. So, changes to how a `Function` is initialized are strictly internal.

Currently, there is no way to initialize a `Function` from "scratch" outside `cudd.pyx`, because `Function.init` is a `cdef`. Suppose that we make the first argument a `BDD` (instead of a `DdManager`). Then, we still cannot call `Function.init` in Python. We must make the second argument a `Function` (instead of a `DdNode`). This second change results also from considering symmetry of arguments to the initializer (`BDD` and `Function` vs `DdManager` and `DdNode`).

We could consider whether it is useful to have the constructor's second argument be a `Function`.

There is little use for such an `init` (which could also be `__init__` after these changes in argument types) in user space. Usually, we make a `Function` to wrap a `DdNode`. So, the reason is that we want to create a `Function` that wraps a lower-level object, not because we want to copy an existing `Function`.

(Python offers other interfaces for copying. Also, I've never needed to copy a `Function`. BDD nodes are treated as immutable objects for almost all purposes. In the context of a specific variable order, this corresponds well to their uniqueness. Therefore, "copying" a `Function` shouldn't be needed for anything, and a second reference to the same `Function` instance should suffice.)

From the above, it follows that an initialization function is needed that takes a `BDD` and a `DdNode`.

**Comment by johnyf** [comment]
Implemented in f30a3c0fe59db8b5c8d341a3b4f9f55a4b8cea86. The reason is to maintain references to the `BDD` instance associated to a `Function` instance, thus preventing Python's garbage collector from deleting the `BDD` (and so the `DdManager`) before the `Function`s associated to that `DdManager`.

aa8788e93006f085585b767ec38c16eabf4f49c1 is a refactoring that avoids repetition of the stanza for initializing `Function` instances. This refactoring is orthogonal to the decision about referencing a `BDD` instead of `DdManager` within `Function`. It would have been nice to decorate the method `Function.init` with `@classmethod`, but Cython [doesn't support that for `cdef` methods](https://groups.google.com/forum/#!topic/cython-users/FLHiLzzKqj4) (and, based on the conclusion from [above](https://github.com/johnyf/dd/issues/13#issue-157128541), `Function.init` must be a `cdef`).

**Comment by johnyf** [comment]
This design decision is leading to several simplifications, and unanticipated solutions to rough edges that have been around for a long time. One reason is that operations at the higher level can be carried out more quickly and automatically (hide some information), by tracking managers of `Function` objects automatically (this is implicit, but that is acceptable in this case, because it is clear in code, and automated from the user's perspective). This *is* flat.

The point is that the principle "flat is better than nested" [cited above](https://github.com/johnyf/dd/issues/13#issue-157128541) had been interpreted in the reverse way than it should. When low-level operations are necessary, *yes*, it is flatter to work directly at the lower level, instead of via attributes from the higher level. But, when the operations involve objects at the higher level, then it is flat to work directly among them (within the higher level), instead of going deep down below, which is nested in that case. These are two complementary notions of "nested". One is the [negative space](https://en.wikipedia.org/wiki/Negative_space) of the other.
